File: Misc/bfs_dfs.cpp
#include<bits/stdc++.h>
using namespace std;
bool dfs(vector<vector<int>> graph, int node, int target, vector<int>& answer){
    answer.push_back(node);
    if (node == target) return true;
    for (auto child : graph[node]){
        if (dfs(graph, child, target, answer)) return true;
    }
    return false;
}

vector<int> bfs(const vector<vector<int>>& graph, int node, int target){
    vector<int> answer;
    vector<bool> visited;
    stack<int> next;
    for (int i = 0; i < graph.size(); i++){
        visited[i] = false;
    }
    visited[node] = true;
    next.push(node);
    while(!next.empty()){
        node = next.top();
        answer.push_back(node);
        cout << node << ' ';
        next.pop();

        for (auto i : graph[node]){
            if (!visited[i]){
                visited[i] = true;
                next.push(i);
            }
        }
    }

    return answer;
}
void bfs2(vector<vector<int>> adj, int s, int target, vector<int>& answer){
    vector<bool> visited(adj.size(), false);
    queue<int> q;
    q.push(s);
    visited[s] = true;

    while (!q.empty()){
        s = q.front();
        q.pop();
        answer.push_back(s);
        if (!visited[target]){
        for (auto i : adj[s]){
            if (!visited[i]){
                q.push(i);
                visited[i] = true;
            }
        }
        }
    }
}

int main(){
    vector<vector<int>> graph = {{1, 2}, {3}, {}, {}};
    for (int i : {0, 1, 2, 3}){
    vector<int> answer = {};
    dfs(graph, 0, i, answer);
    for (auto p : answer) cout << p << ' ';
    cout << endl;
    }

    for (int i : {0, 1, 2, 3}){
    vector<int> answer = {};
    bfs2(graph, 0, i, answer);
    for (auto p : answer) cout << p << ' ';
    cout << endl;
    }

   
    

}

File: Misc/bfs_dfs.py
from collections import deque

def dfs(graph, node, target, output_list):
    output_list.append(node)
    if node == target:
        return True
    children = graph[node]
    for child in children:
        if dfs(graph, child, target, output_list):
            return True
    return False

def run_dfs(graph, node, target):
    output_list = []
    dfs(graph, node, target, output_list)
    return output_list

def bfs(graph, node, target):
    queue = deque()
    output_list = []
    visited = [False for pair in list(graph.items())]
    visited[node - 1] = True
    queue.append(node)
    while len(queue):
        output_list.append(queue.popleft())
        for child in graph[node]:
            if(not visited[child - 1]):
                visited[child - 1] = True
                queue.append(child)

    return output_list

def main():    
    graph = {1: [2, 3], 2: [4], 3: [], 4: []}
    dfs(graph, 1, 4)
    bfs(graph, 1, 4)

if __name__ == '__main__':
    main()